### ISSUE TITLE
[BUG] python orc/parquet writer test failures in CI for 21.12

### ISSUE THREAD

**Issue by abellina** [rapidsai/cudf#9408]
I am seeing the following tests failing in one of my PRs:

cudf.tests.test_gcs.test_write_orc
cudf.tests.test_s3.test_write_parquet
cudf.tests.test_s3.test_write_orc
cudf.tests.test_orc.test_orc_bool_encode_fail
cudf.tests.test_orc.test_nanoseconds_overflow
cudf.tests.test_orc.test_empty_dataframe
cudf.tests.test_parquet.test_parquet_write_bytes_io
cudf.tests.test_parquet.test_parquet_writer_bytes_io
cudf.tests.test_parquet.test_parquet_write_partitioned
cudf.tests.test_parquet.test_parquet_write_to_dataset
cudf.tests.test_parquet.test_parquet_index
cudf.tests.test_orc.test_empty_string_columns
cudf.tests.test_orc.test_orc_writer_lists
cudf.tests.test_orc.test_writer_lists_structs
cudf.tests.test_orc.test_orc_string_stream_offset_issue
cudf.tests.test_orc.test_orc_writer_lists_empty_rg
cudf.tests.test_orc.test_chunked_orc_writer_lists

This issue seems related: https://github.com/rapidsai/cudf/pull/9372 fyi @devavret. 

Here's the output for one of them:
```
tmpdir_factory = TempdirFactory(_tmppath_factory=TempPathFactory(_given_basetemp=PosixPath('/workspace/cudf-cuda-tmp/popen-gw10'), _trace=<pluggy._tracing.TagTracerSub object at 0x7fda2be2c750>, _basetemp=PosixPath('/workspace/cudf-cuda-tmp/popen-gw10')))
cols = ['b'], filename = None

    @pytest.mark.parametrize("filename", ["myfile.parquet", None])
    @pytest.mark.parametrize("cols", [["b"], ["c", "b"]])
    def test_parquet_write_partitioned(tmpdir_factory, cols, filename):
        # Checks that write_to_dataset is wrapping to_parquet
        # as expected
        gdf_dir = str(tmpdir_factory.mktemp("gdf_dir"))
        pdf_dir = str(tmpdir_factory.mktemp("pdf_dir"))
        size = 100
        pdf = pd.DataFrame(
            {
                "a": np.arange(0, stop=size, dtype="int64"),
                "b": np.random.choice(list("abcd"), size=size),
                "c": np.random.choice(np.arange(4), size=size),
            }
        )
        pdf.to_parquet(pdf_dir, index=False, partition_cols=cols)
        gdf = cudf.from_pandas(pdf)
        gdf.to_parquet(
>           gdf_dir, index=False, partition_cols=cols, partition_file_name=filename
        )

cudf/tests/test_parquet.py:1560: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/opt/conda/envs/rapids/lib/python3.7/site-packages/cudf/core/dataframe.py:6669: in to_parquet
    return pq.to_parquet(self, path, *args, **kwargs)
/opt/conda/envs/rapids/lib/python3.7/site-packages/cudf/io/parquet.py:435: in to_parquet
    **kwargs,
/opt/conda/envs/rapids/lib/python3.7/site-packages/cudf/io/parquet.py:127: in write_to_dataset
    write_df.to_parquet(fil, index=preserve_index, **kwargs)
/opt/conda/envs/rapids/lib/python3.7/site-packages/cudf/core/dataframe.py:6669: in to_parquet
    return pq.to_parquet(self, path, *args, **kwargs)
/opt/conda/envs/rapids/lib/python3.7/site-packages/cudf/io/parquet.py:470: in to_parquet
    int96_timestamps=int96_timestamps,
cudf/_lib/parquet.pyx:264: in cudf._lib.parquet.write_parquet
    cpdef write_parquet(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   out_metadata_c = move(parquet_writer(args))
E   RuntimeError: cuDF failure at: /workspace/cpp/include/cudf/io/data_sink.hpp:161: data_sink classes that support device_write_async must override it.

cudf/_lib/parquet.pyx:341: RuntimeError
```

**Comment by vuule** [comment]
Looks like this is exactly the same error we made with the async reads :facepalm:

**Comment by firestarman** [comment]
I got similar errors in java tests in my PR https://github.com/rapidsai/cudf/pull/9334.
```
[ERROR] testORCWriteToBufferChunked  Time elapsed: 0.005 s  <<< ERROR!
ai.rapids.cudf.CudfException: cuDF failure at: /opt/conda/envs/rapids/include/cudf/io/data_sink.hpp:161: data_sink classes that support device_write_async must override it.
	at ai.rapids.cudf.Table.writeORCChunk(Native Method)
	at ai.rapids.cudf.Table.access$1000(Table.java:39)
```

**Comment by devavret** [comment]
>Can we make the implementation of `device_write_async` be optional if it is an improvement ?

That's a good idea. But I'd still like to figure out the failures. This time, I added `device_write_async` to every class that had a `device_write` method. So what's causing the error?

**Comment by vuule** [comment]
Are the affected tests perhaps using custom data sink types?
If that's the issue, we can have the async write call the sync version so custom sink implementations don't have to implement their own versions.

**Comment by devavret** [comment]
> My PR (#9406) was missing #9372, which makes it weird I'd hit this `CUDF_FAIL`, since I didn't have the `device_write_async` change in my branch. I upmerged and it's rebuilding now.

This is weird. IDK what CI does when the PR branch is behind the mainline. Can headers be updated in CI without merging the latest branch? Also, I cannot find any mention of cuFile in the build log. @vuule Is GDS tested in CI?

**Comment by firestarman** [comment]
> Update: my build for #9406 passed on the python side of things, so adding the commit fixed it (as suspected?) Now the java CI is broken as per @firestarman's comment. I'll resolve that next.

@abellina I have got a quick fix in my PR #9334, could you help review it?

**Comment by abellina** [comment]
So lets close this particular issue, and follow up on #9430  @devavret. If you have an issue on the C++ side please link in.
